### PR TITLE
Mac: Support Dark Mode in Advanced View

### LINKED ISSUE #3115: [Manager] Support dark theme mode on Linux/MacOS
Currently BOINC Manager is almost unusable when using dark theme on Linux.
It is a consolidated ticket that includes next issues:

- #812
- #980
- #1072
- #1126
- #2919
- #3112 
- #3415
- #4023

### PR DESCRIPTION
**Pull request by CharlieFenton**
Fixes #3115 (at least for macOS)

Implementing Dark Mode support for macOS and possibly for Windows and Linux. I am initially not guarding my changes with `#ifdef __APPLE__ ` or `#ifdef __WXMAC__` to allow testing of theWindows and Linux CI build artifacts to see how well my Mc changes work on those platforms. I expect to add those guards before submitting this PR for merging.

So far, I have Dark Mode working on the Projects, Tasks, Transfers and Disk tabs in Advanced View, as well as the Event Log and various Advanced View dialogs. I have not been able to work on the Notices tab because my attached BOINC projects currently have no notices.

I still need to implement Dark Mode for the Simple View.

### CODE REVIEW

**Comment by AenBleidd** [comment]
@CharlieFenton, thank you for this implementation.
if you don't mind, please don't rush with those guards. There is a high chance that it might work on Linux and Windows as well.
I'll check this week om Windows and linux, and will try to provide fixes for aby possible UI issues.

**Comment by CharlieFenton** [comment]
@AenBleidd Please let me finish it on the Mac before you get too deeply into texting on other platforms. If this can be made to work on other platforms with a reasonable number of fixes, I would prefer that be done in a separate PR after this one has been merged.

**Comment by Vulpine05** [comment]
> So far, I have Dark Mode working on the Projects, Tasks, Transfers and Disk tabs in Advanced View, as well as the Event Log and various Advanced View dialogs. I have not been able to work on the Notices tab because my attached BOINC projects currently have no notices.
> 
> I still need to implement Dark Mode for the Simple View.

@CharlieFenton, if you have an app_config file for a project, just make a typo under one of the headers.  That should create a notice.

**Comment by AenBleidd** [comment]
I made a quick test on Windows.
Since the latest released version of wxWidgets we are using to build BOINC doesn't support dark mode, only the self-drawn elements change their color.
For now dark mode should be disabled for Windows.
I'll keep and eye on wxWidgets, and when they will implement this feature, I'll give another try.
Later will check also the Linux version.
Hopefully at least there is will work fine.

**Comment by CharlieFenton** [comment]
@AenBleidd Thank you for testing. Once I hear your Linux results, I will add guards to the code so it does nothing on Windows, and perhaps also Linux if needed.

File: clientgui/AdvancedFrame.cpp
Comment (Vulpine05) [inline clientgui/AdvancedFrame.cpp:1984]: Add a blank comment line after this line.

File: clientgui/AdvancedFrame.cpp
Comment (Vulpine05) [inline clientgui/AdvancedFrame.cpp:1994]: Spell check:  "control's"

File: clientgui/AdvancedFrame.cpp
Comment (Vulpine05) [inline clientgui/AdvancedFrame.cpp:2004]: Add a blank comment line after this line.

File: clientgui/NoticeListCtrl.cpp
Comment (Vulpine05) [inline clientgui/NoticeListCtrl.cpp:113]: Spell check:  window

**Comment by CharlieFenton** [comment]
@AenBleidd Dark Mode in Advanced View on Mac is now fully working; I have not finished implementing it for Simple View. Please test Dark Mode Advanced View on Linux today if possible.

**Comment by AenBleidd** [comment]
@CharlieFenton, sure, will try to do that. I'll let you know when I get the results

**Comment by CharlieFenton** [comment]
@AenBleidd I have set it up so that the selection of which platforms use the Dark Mode code requires changes in only one line: _clientgui/BOINCGUIApp.h_ line 44. I have temporarily set that line to allow both MacOS and Linux:
`#if (defined(__WXMAC__) || defined(__WXGTK__))`

Again, Advanced Mode is fully implemented for macOS, but  I am still working on changes for Simple Mode. Please test Advanced Mode on Linux.

**Comment by CharlieFenton** [comment]
@AenBleidd I think I have now completed implementing Dark Mode for the Macintosh. The only thing left is for me to remove `|| defined(__WXGTK__)` from line 44 of _clientgui/BOINCGUIApp.h_ unless you find that it works on Linux. Please test this as soon as possible and let me know so I can change this from draft.

**Comment by CharlieFenton** [comment]
@AenBleidd I just realized I had not yet checked in the` || defined(__WXGTK__)` to enable the Dark mode code on Linux. I have fixed that now.

**Comment by 47an** [comment]
Linux greger-ASUS-TUF 5.19.0-43-generic #44~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon May 22 13:39:36 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Screenshot from 2023-06-16 23-02-36
Screenshot from 2023-06-16 23-13-03
Screenshot from 2023-06-16 23-33-41
Screenshot from 2023-06-16 23-37-01
Screenshot from 2023-06-16 23-37-08
Screenshot from 2023-06-16 23-37-15
Screenshot from 2023-06-16 23-37-29
Screenshot from 2023-06-16 23-37-35
Screenshot from 2023-06-16 23-37-41

**Comment by AenBleidd** [comment]
@CharlieFenton, I failed to test this, so I asked for a help with this PR.
As you can see above, dark mode works fine on Linux (at least, I see no issues with it).
So I may assume, it's safe to enable dark mode for OSX and Linux.
For Windows we still need to wait until this feature will be fully implemented and released be wxWidgets team (current 3.2 release doesn't include it), and I created a separate ticket for it: #5274.
I'll keep an eye on it and will test once available.
Also, later today I'll try to review this PR.

**Comment by KeithMyers** [comment]
Here is the only page that isn't in Dark mode.

Screenshot from 2023-06-16 14-52-51

**Comment by AenBleidd** [comment]
'Notices' is the web page. Not sure it's reasonable to convert it to dark mode manually.
For me it's completely fine to have it as-is

**Comment by KeithMyers** [comment]
Just wanted to also provide some error messages from this boincmgr version when executed.

 ./boincmgr
Gtk-Message: 14:52:09.356: Failed to load module "appmenu-gtk-module"
Gtk-Message: 14:52:09.391: Failed to load module "canberra-gtk-module"
Gtk-Message: 14:52:09.393: Failed to load module "canberra-gtk-module"

Already have both of the supposed missing libraries loaded.

$ apt search appmenu-gtk-module
Sorting... Done
Full Text Search... Done
appmenu-gtk-module-common/jammy,jammy,now 0.7.6-2 all [installed,automatic]
  Common files for GtkMenuShell D-Bus exporter

$ apt search libcanberra-gtk-module
Sorting... Done
Full Text Search... Done
libcanberra-gtk-module/jammy-updates,now 0.30-10ubuntu1.22.04.1 amd64 [installed]
  translates GTK+ widgets signals to event sounds

**Comment by KeithMyers** [comment]
> 'Notices' is the web page. Not sure it's reasonable to convert it to dark mode manually. For me it's completely fine to have it as-is

Ok, didn't realize that Notices was an actual web page pulled from the projects.

**Comment by davidpanderson** [comment]
Notices is part of the manager.  It's displayed using an HTML renderer that's part of WxWidgets.  In dark mode it would ideally be dark too.

**Comment by AenBleidd** [comment]
@davidpanderson, not sure wxWidgets 'smart enough' to convert colors on the html page.
In any case, I'd say that this is not something that should be fixed on our side, since it might require quite a huge effort (I'll check if this component of wxWidgets supports dark mode at all, and probably will talk to the wxWidgets maintainers about it)

**Comment by CharlieFenton** [comment]
On the Mac, Notices uses _wxWebView_ and my code does shift its display to whit text on black in Dark Mode.  On at least some Linux versions, Notices uses _wxHtmlWindow_ because _wxWebView_ is unavailable or doesn't work. I h ave no way to test on Linux, but I will see  if I can get it working with _wxHtmlWindow_ by **temporarily** modifying the Mac code to use _wxHtmlWindow_.

**Comment by CharlieFenton** [comment]
@AenBleidd I noticed one other small thing I need to fix on the Mac before  it is ready to merge. I wi ll also try to get it working with _wxHtmlWindow_ for the Notices.

**Comment by CharlieFenton** [comment]
Here is what the Notices tab looks like on the Mac in Dark mode:

Screenshot 2023-06-16 at 5 07 54 PM

**Comment by CharlieFenton** [comment]
@KeithMyers I believe I have fixed Dark Mode for the Notices tab under Advanced View and Notices under Simple View in Linux. Please test it and let us know the result. Also, please check the various dialogs (Select Computer, Computing Preferences (both Simple and Advanced Views), Other Options, Event Log Options, Exclusive Applications, Select Columns, Add Project / Account Manager, etc.)

**Comment by KeithMyers** [comment]
New artifact works on all tabs of the Manager now.

Screenshot from 2023-06-18 06-56-56
Screenshot from 2023-06-18 07-00-23
Screenshot from 2023-0
![Screenshot from 2023-06-18 07-02-46
Screenshot from 2023-06-18 07-05-50
Screenshot from 2023-06-18 07-06-49

**Comment by KeithMyers** [comment]
Screenshot from 2023-06-18 07-12-17
Screenshot from 2023-06-18 07-11-39
Screenshot from 2023-06-18 07-02-46
Screenshot from 2023-06-18 07-01-02
Screenshot from 2023-06-18 07-00-23
Screenshot from 2023-06-18 06-56-56

**Comment by AenBleidd** [comment]
Looks awesome. Thanks, @CharlieFenton, @KeithMyers!
I'll make a review tonight then.

**Comment by davidpanderson** [comment]
Great work!  Small suggestion:
in the striped areas (tasks, projects) make the light gray darker.

**Comment by KeithMyers** [comment]
> Great work! Small suggestion: in the striped areas (tasks, projects) make the light gray darker.

Not too much though.  I think it's fine in my opinion.

**Review by AenBleidd — APPROVED** [review]
Looks good to me

**Comment by KeithMyers** [comment]
> 

I too am fine with the color choices as is.  Niggles only for any commentary from other users.

**Comment by CharlieFenton** [comment]
Our code is not setting the colors and doesn't have control over the choices. wxWidgets sets the Light Mode colors and the OS (or perhaps code in wxWidgets) decides what to change them to for Dark Mode. All we do to create the alternating row colors (which wxWidgets calls "Zebra striping") is call `EnableAlternateRowColours(true);` .

**Comment by CharlieFenton** [comment]
> I've gone through all the menus and selected various options. All choices give me Dark Mode view.

@KeithMyers Does this mean that you were able to confirm that the dialogs I listed [here](https://github.com/BOINC/boinc/pull/5267#issuecomment-1596131354) also display properly in Dark Modern Linux? If so, this PR is ready to be merged.

**Comment by CharlieFenton** [comment]
@KeithMyers Thank you for your help with this.

**Comment by KeithMyers** [comment]
> > I've gone through all the menus and selected various options. All choices give me Dark Mode view.
> 
> @KeithMyers Does this mean that you were able to confirm that the dialogs I listed [here](https://github.com/BOINC/boinc/pull/5267#issuecomment-1596131354) also display properly in Dark Modern Linux? If so, this PR is ready to be merged.

Yes, I went through every choice you mentioned and confirmed they all were in Dark Mode.

**Pull request merged**
